### PR TITLE
B0 client

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Refactor the bl_crypto_client into bl_crypto.

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
cc @hakonfam @sigvartmh

**Comment by thst-nordic** [comment]
We have enabled new documentation checks in the CI  
If you get a new failure in 'CI/Jenkins/DOC', try rebasing from master.  
If this doesn't fix things or you have other CI questions,  you can get help in the Teams chatroom 'NCS->NCS CI'

File: samples/nrf9160/secure_services/src/main.c [hunk 1]
```diff
@@ -63,6 +63,10 @@ void main(void)
 	}
 
 	print_hex_number(buf, num_bytes_to_read);
+#elif defined(PM_S0_ADDRESS)
+	bool valid = spm_prevalidate_b1_upgrade(PM_S0_ADDRESS, PM_S0_ADDRESS);
```
Comment (hakonfam) [inline samples/nrf9160/secure_services/src/main.c]: Why is the same address given twice?

Comment (oyvindronningstad) [reply samples/nrf9160/secure_services/src/main.c]: Because the function takes two addresses, current and target, but here we call the function on the current b1 for demonstration purposes.

**Comment by hakonfam** [comment]
All dependencies are now merged

File: subsys/bootloader/Kconfig [hunk 2]
```diff
@@ -12,13 +12,14 @@ config MCUBOOT_BUILD_S1_VARIANT
 	help
 	  Build upgrade candidate of MCUBoot for alternative slot S1.
 
-config SECURE_BOOT_CRYPTO_CLIENT
-	bool "Include Secure Bootloader crypto client API"
+config SECURE_BOOT_CLIENT_API
+	bool "Include Secure Bootloader client API"
+	select FW_METADATA
```
Comment (hakonfam) [inline subsys/bootloader/Kconfig]: This should be "depends on" 

From the zephyr docs: `Avoid selecting symbols with prompts or dependencies. Prefer depends on`

File: subsys/bootloader/bl_client/CMakeLists.txt [hunk 3]
```diff
@@ -5,6 +5,3 @@
 #
 zephyr_library()
 zephyr_library_sources(bl_client.c)
-zephyr_link_libraries(nrfxlib_crypto)
```
Comment (hakonfam) [inline subsys/bootloader/bl_client/CMakeLists.txt]: This change is not explained in the commit message. Why are these removed?

File: subsys/bootloader/bl_client/bl_client.c [hunk 4]
```diff
@@ -59,6 +59,20 @@ static const struct bl_secp256r1_abi *get_bl_secp256r1_abi(void)
 	return (const struct bl_secp256r1_abi *)abi;
 }
 
+static const struct bl_validate_fw_abi *get_bl_validate_firmware_abi(void)
+{
+	static const struct fw_abi_info *abi;
+
+	if (!abi) {
+		abi = fw_abi_find(BL_VALIDATE_FW_ABI_ID,
+				CONFIG_BL_VALIDATE_FW_ABI_FLAGS,
```
Comment (hakonfam) [inline subsys/bootloader/bl_client/bl_client.c]: Are you missing the patch which adds these kconfigs? I don't see them defined in this PR.

File: subsys/bootloader/bl_client/bl_client.c [hunk 5]
```diff
@@ -97,3 +111,9 @@ int bl_secp256r1_validate(const u8_t *hash, u32_t hash_len,
 	return get_bl_secp256r1_abi()->abi.bl_secp256r1_validate(hash, hash_len,
 							public_key, signature);
 }
+
+bool bl_validate_firmware(u32_t fw_dst_address, u32_t fw_src_address)
```
Comment (hakonfam) [inline subsys/bootloader/bl_client/bl_client.c]: Where is this function linked in to the bl_client ABI?

File: subsys/bootloader/bl_client/bl_client.c (hunk 4 above)
Comment (hakonfam) [reply subsys/bootloader/bl_client/bl_client.c]: I get a compile error because of these missing.

Comment (sigvartmh) [reply subsys/bootloader/bl_client/bl_client.c]: I'm pretty sure they are located in the FW_METADATA template PR but i'm not 100% certain.

File: subsys/bootloader/bl_client/CMakeLists.txt (hunk 3 above)
Comment (oyvindronningstad) [reply subsys/bootloader/bl_client/CMakeLists.txt]: They were needed earlier for #include files, but that has been fixed/changed a while ago, but these weren't removed then. I'll see if I can mention it in the commit message.

File: subsys/bootloader/bl_client/bl_client.c (hunk 5 above)
Comment (oyvindronningstad) [reply subsys/bootloader/bl_client/bl_client.c]: I'm not entirely sure what you are asking, but as a test you could read the newly written fw_metadata docs and see if you understand it afterwards :).

fw_metadata docs are in this PR: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1291

Comment (oyvindronningstad) [reply subsys/bootloader/bl_client/bl_client.c]: Or comment there what questions it doesn't answer.

File: subsys/bootloader/bl_client/CMakeLists.txt (hunk 3 above)
Comment (oyvindronningstad) [reply subsys/bootloader/bl_client/CMakeLists.txt]: Added some explanation in the commit message.

File: subsys/bootloader/Kconfig (hunk 2 above)
Comment (oyvindronningstad) [reply subsys/bootloader/Kconfig]: Done

**Comment by oyvindronningstad** [comment]
DNM until #1378 is decided.

**Comment by oyvindronningstad** [comment]
Closing since the original solution was dropped.
